client/post-editor/editor-location/index.jsx

```diff
@@ -48,23 +49,71 @@ class EditorLocation extends React.Component {
 				return new Error( 'Expected array pair of coordinates for prop `' + propName + '`.' );
 			}
 		},
+		isSharedPublicly: PropTypes.bool,
 	};
 
 	state = {
 		error: null,
+		previouslyPrivatePostBeingModified: false,
 	};
 
+	constructor( props ) {
+		super( props );
+
+		this.originalProps = props;
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		// When geo data was originally private and its coordinates have changed, we track that change
+		// so we can warn the user that after saving the information will be publicly visible.
+		if (
+			'private' === this.originalProps.isSharedPublicly &&
+			( this.originalProps.coordinates && nextProps.coordinates ) &&
+			( this.originalProps.coordinates[ 0 ] !== nextProps.coordinates[ 0 ] ||
+				this.originalProps.coordinates[ 1 ] !== nextProps.coordinates[ 1 ] )
+		) {
+			this.setState( { previouslyPrivatePostBeingModified: true } );
+		}
+	}
+
 	onGeolocateSuccess = position => {
-		this.setState( {
-			locating: false,
-		} );
+		const latitude = toGeoString( position.coords.latitude ),
+			longitude = toGeoString( position.coords.longitude );
 
 		this.props.updatePostMetadata( this.props.siteId, this.props.postId, {
-			geo_latitude: toGeoString( position.coords.latitude ),
-			geo_longitude: toGeoString( position.coords.longitude ),
+			geo_latitude: latitude,
+			geo_longitude: longitude,
+			geo_public: 1,
 		} );
 
 		recordStat( 'location_geolocate_success' );
+
+		reverseGeocode( latitude, longitude )
+			.then( results => {
+				const localityResults = results.filter( result => {
+					return -1 !== result.types.indexOf( 'locality' );
+				} );
+
+				if ( localityResults.length ) {
+					this.props.updatePostMetadata( this.props.siteId, this.props.postId, {
+						geo_address: localityResults[ 0 ].formatted_address,
+					} );
+				}
+
+				recordStat( 'location_reverse_geocode_success' );
+			} )
+			.catch( () => {
+				this.props.updatePostMetadata( this.props.siteId, this.props.postId, {
+					geo_address: latitude + ', ' + longitude,
+				} );
+
+				recordStat( 'location_reverse_geocode_failed' );
+			} )
+			.finally( () => {
```

Comment:
There are polyfills available in Babel but a second then() actually has the same effect in this case.  I switched to that approach in b012eee to avoid potential compatibility issues.  then() is maybe marginally less helpful versus finally() semantically in this case, but I think it's a reasonable trade-off.  (Here is the actual MDN compatibility table for the feature, by the way: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally#Browser_compatibility)